src/imperative.rkt

```diff
@@ -98,142 +98,167 @@
           (printf "~a\t~a\n" indent (convert-assignment var* var**)))
-      (printf "~a\t~a\n" indent
-          (convert-assignment test-var (convert-expr cond #:ctx ctx** #:indent indent*)))
+
+      (define-values (cond** cond**-prec) (convert-expr cond #:ctx ctx** #:indent indent*))
+      (printf "~a\t~a\n" indent (convert-assignment test-var cond**))
       (printf "~a}~a\n" indent (if (equal? (language-name (*lang*)) "sollya") ";" ""))
       (convert-expr retexpr #:ctx ctx* #:indent indent)]
 
     [`(while* ,cond ([,vars ,inits ,updates] ...) ,retexpr)
       (define indent* (format "~a\t" indent))
       (define-values (ctx* vars*)
         (for/fold ([ctx* ctx] [vars* '()]) ([var vars] [val inits])
-          (let-values ([(cx name) (ctx-unique-name ctx* var)])
-            (printf "~a~a\n" indent
-                (convert-declaration cx name (convert-expr val #:ctx ctx* #:indent indent)))
+          (let*-values ([(val* prec) (convert-expr val #:ctx ctx* #:indent indent)]
+                        [(cx name) (ctx-unique-name ctx* var prec)]
+                        [(cx*) (ctx-update-props cx `(:precision ,prec))])
+            (printf "~a~a\n" indent (convert-declaration cx* name val*))
             (values cx (flatten (cons vars* name))))))
       (define test-var
         (let-values ([(cx name) (ctx-random-name ctx)])
             (set! ctx cx)
             name))
       (printf "~a" (use-vars vars* indent))
+
+      (define-values (cond* cond*-prec) (convert-expr cond #:ctx ctx* #:indent indent))
       (printf "~a~a\n" indent
-          (convert-declaration
-              (ctx-update-props ctx '(:precision boolean))
-              test-var
-              (convert-expr cond #:ctx ctx* #:indent indent)))
+          (convert-declaration (ctx-update-props ctx '(:precision boolean))
+                               test-var cond*))
       ; Sollya has slightly different while*
       (if (equal? (language-name (*lang*)) "sollya")
           (printf "~awhile ~a do {\n" indent (format-cond test-var))
           (printf "~a~a ~a {\n" indent (while-name) (format-cond test-var)))
       (for ([var* vars*] [update updates])
-        (printf "~a\t~a\n" indent
-            (convert-assignment var* (convert-expr update #:ctx ctx* #:indent indent*))))
-      (printf "~a\t~a\n" indent
-          (convert-assignment test-var (convert-expr cond #:ctx ctx* #:indent indent*)))
-       (printf "~a}~a\n" indent (if (equal? (language-name (*lang*)) "sollya") ";" ""))
+        (let-values ([(val* _) (convert-expr update #:ctx ctx* #:indent indent*)])
+          (printf "~a\t~a\n" indent (convert-assignment var* val*))))
+
+      (define-values (cond** cond**-prec) (convert-expr cond #:ctx ctx* #:indent indent*))
+      (printf "~a\t~a\n" indent (convert-assignment test-var cond**))
+      (printf "~a}~a\n" indent (if (equal? (language-name (*lang*)) "sollya") ";" ""))
       (convert-expr retexpr #:ctx ctx* #:indent indent)]
 
     ; Ignore all casts
```

Comment:
I think this comment ("Ignore all casts") should be removed because casts are not ignored. This comment should be removed from core2smtlib2.rkt also.

Reply:
Oops. Those are obsolete comments.